src/OpenTK.Windowing.GraphicsLibraryFramework/JoystickState.cs

```diff
@@ -0,0 +1,197 @@
+        {
+            return _hats[_current][index];
+        }
+
+        /// <inheritdoc />
+        public Hat GetHatPrevious(int index)
+        {
+            return _hats[_previous][index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetHat(int index, Hat value)
+        {
+            _hats[_current][index] = value;
+        }
+
+        /// <inheritdoc />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool IsButtonDown(int index)
+        {
+            byte b = _buttons[_current][index / 8];
+            int pow = (int)Math.Pow(2, index % 8);
+
+            return (b & pow) == pow;
+        }
+
+        /// <inheritdoc />
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool WasButtonDown(int index)
+        {
+            byte bCurrent = _buttons[_current][index / 8];
+            byte bPrevious = _buttons[_previous][index / 8];
+            int pow = (int)Math.Pow(2, index % 8);
+
+            return (bCurrent & pow) != pow && (bPrevious & pow) == pow;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetButtonDown(int index, bool value)
+        {
+            int byteOffSet = index / 8;
+            int bitOffset = index % 8;
+
+            if (value)
+            {
+                _buttons[_current][byteOffSet] |= (byte)(1 << bitOffset);
+            }
+            else
+            {
+                _buttons[_current][byteOffSet] &= (byte)~(1 << bitOffset);
+            }
+        }
+
+        /// <inheritdoc />
+        public float GetAxis(int index)
+        {
+            return _axes[_current][index];
+        }
+
+        /// <inheritdoc />
+        public float GetAxisPrevious(int index)
+        {
+            return _axes[_previous][index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetAxis(int index, float value)
+        {
+            _axes[_current][index] = value < -1 ? -1 : (value > 1 ? 1 : value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetAxes(float[] axes)
+        {
+            _axes[_current] = axes;
+        }
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            // string.Join internally uses a StringBuilder
+            var joinedHats = string.Join(", ", _hats[_current]);
+            var joinedAxes = string.Join(", ", _axes[_current]);
+
+            var buttonBuilder = new StringBuilder();
+
+            for (int i = 0; i < _buttons.Length * 8; i++)
+            {
+                buttonBuilder.Append($", {(IsButtonDown(i) ? "down" : "up")}");
+            }
+
+            // String Interpolation uses string.Format which internally uses a StringBuilder
+            return $"{{hats: [{joinedHats}], axes: [{joinedAxes}], buttons: [{buttonBuilder}], id: {Id}, name: {Name}}}";
```

Comment:
I think `Id` and `Name` should be first, as that probably tells you more at a glance, and the rest of the string can probably get pretty long.

Reply:
We can do that. It differs from the original implementation, but I agree it's a good idea.